File: frontend/src/shared/types/canvas.ts
export type NodeStatus = 'idle' | 'queued' | 'running' | 'done' | 'failed'

export type CanvasNodeType =
  | 'text'
  | 'image'
  | 'video'
  | 'audio'
  | 'note'
  | 'task'
  | 'text-gen'
  | 'image-gen'
  | 'video-gen'
  | 'asset'
  | 'review'
  | 'decision'
  | 'template-ref'

export interface CanvasPosition {
  x: number
  y: number
}

export interface CanvasNode {
  id: string
  type: CanvasNodeType
  title: string
  description?: string
  prompt?: string
  content?: string | null
  position: CanvasPosition
  status: NodeStatus
  model?: string
  width?: number
  height?: number
  badge?: string
  sourceKind?: 'system' | 'upload'
  runCount?: number
  lastRunAt?: string | null
  settings?: Record<string, string>
  outputText?: string | null
  errorMessage?: string | null
  panelColor?: string | null
}

export interface CanvasEdge {
  id: string
  source: string
  target: string
}

export interface CanvasGroup {
  id: string
  title: string
  nodeIds: string[]
  color?: string
}

export interface CanvasViewport {
  x: number
  y: number
  zoom: number
}

export interface CanvasSnapshot {
  version: number
  savedAt: string
  viewport: CanvasViewport
  nodes: CanvasNode[]
  edges: CanvasEdge[]
  groups: CanvasGroup[]
}

File: backend/app/schemas/canvas.py
from typing import Literal

from pydantic import BaseModel, Field

NodeStatus = Literal["idle", "queued", "running", "done", "failed"]
CanvasNodeType = Literal[
    "note",
    "task",
    "text-gen",
    "image-gen",
    "video-gen",
    "asset",
    "review",
    "decision",
    "template-ref",
]


class CanvasPosition(BaseModel):
    x: float
    y: float


class CanvasNode(BaseModel):
    id: str
    type: CanvasNodeType
    title: str
    description: str | None = None
    position: CanvasPosition
    status: NodeStatus = "idle"
    model: str | None = None


class CanvasEdge(BaseModel):
    id: str
    source: str
    target: str


class CanvasGroup(BaseModel):
    id: str
    title: str
    node_ids: list[str] = Field(default_factory=list, alias="nodeIds")


class CanvasViewport(BaseModel):
    x: float = 0
    y: float = 0
    zoom: float = 1


class CanvasSnapshot(BaseModel):
    version: int = 1
    saved_at: str = Field(alias="savedAt")
    viewport: CanvasViewport
    nodes: list[CanvasNode] = Field(default_factory=list)
    edges: list[CanvasEdge] = Field(default_factory=list)
    groups: list[CanvasGroup] = Field(default_factory=list)
